Mathlib/Tactic/Module.lean

```diff
@@ -0,0 +1,643 @@
+  -- now run `simp` with these lemmas, and (importantly) *no* simprocs
+  let ctx : Simp.Context := {
+      config      := { failIfUnchanged := false }
+      simpTheorems := #[thms]
+    }
+  let (some r, _) ← simpTarget mvarId ctx (simprocs := #[]) |
+    throwError "internal error in match_scalars tactic: postprocessing should not close goals"
+  return r
+
+/-- Given a goal which is an equality in a type `M` (with `M` an `AddCommMonoid`), parse the LHS and
+RHS of the goal as linear combinations of `M`-atoms over some semiring `R`, and reduce the goal to
+the respective equalities of the `R`-coefficients of each atom. -/
+def matchScalars (g : MVarId) : MetaM (List MVarId) := do
+  let mvars ← AtomM.run .instances (matchScalarsAux g)
+  mvars.mapM postprocess
+
+/-- Given a goal which is an equality in a type `M` (with `M` an `AddCommMonoid`), parse the LHS and
+RHS of the goal as linear combinations of `M`-atoms over some semiring `R`, and reduce the goal to
+the respective equalities of the `R`-coefficients of each atom.
+
+For example, this produces the goal `⊢ a * 1 + b * 1 = (b + a) * 1`:
+```
+example [AddCommMonoid M] [Semiring R] [Module R M] (a b : R) (x : M) :
+    a • x + b • x = (b + a) • x := by
+  match_scalars
+```
+This produces the two goals `⊢ a * (a * 1) + b * (b * 1) = 1` (from the `x` atom) and
+`⊢ a * -(b * 1) + b * (a * 1) = 0` (from the `y` atom):
+```
+example [AddCommGroup M] [Ring R] [Module R M] (a b : R) (x : M) :
+    a • (a • x - b • y) + (b • a • y + b • b • x) = x := by
+  match_scalars
+```
+This produces the goal `⊢ -2 * (a * 1) = a * (-2 * 1)`:
+```
+example [AddCommGroup M] [Ring R] [Module R M] (a : R) (x : M) :
+    -(2:R) • a • x = a • (-2:ℤ) • x  := by
+  match_scalars
+```
+-/
+elab "match_scalars" : tactic => Tactic.liftMetaTactic matchScalars
+
+/-- Given a goal which is an equality in a type `M` (with `M` an `AddCommMonoid`), parse the LHS and
+RHS of the goal as linear combinations of `M`-atoms over some commutative semiring `R`, and prove
+the goal by checking that the LHS- and RHS-coefficients of each atom are the same up to
+ring-normalization in `R`.
+
+Examples:
+```
+example [AddCommMonoid M] [CommSemiring R] [Module R M] (a b : R) (x : M) :
+    a • x + b • x = (b + a) • x := by
+  module
+
+example [AddCommMonoid M] [Field K] [CharZero K] [Module K M] (x : M) :
+    (2:K)⁻¹ • x + (3:K)⁻¹ • x + (6:K)⁻¹ • x = x := by
+  module
+
+example [AddCommGroup M] [CommRing R] [Module R M] (a : R) (v w : M) :
+    (1 + a ^ 2) • (v + w) - a • (a • v - w) = v + (1 + a + a ^ 2) • w := by
+  module
+
+example [AddCommGroup M] [CommRing R] [Module R M] (a b μ ν : R) (x y : M) :
+    (μ - ν) • a • x = (a • μ • x + b • ν • y) - ν • (a • x + b • y) := by
+  module
+```
+
+If the proofs of coefficient-wise equality will require more reasoning than just ring-normalization,
+use the tactic `match_scalars` instead, and then prove coefficient-wise equality by hand.
```

Comment:
Personally, I would put this comment *before* the examples, just so that it's easier for the user to find.